### ISSUE TITLE
How to refresh updated Claims without login out

### ISSUE THREAD

**Issue by Alienroid** [dotnet/AspNetCore.Docs#22405]
I am using Blazor Server Side.

This document specified "SignInManager<TUser> and UserManager<TUser> aren't supported in Razor components."

I have a custom Claim (example FirstName) by following the documentation. I can read it with "user.Claims.FirstOrDefault(s => s.Type == "FirstName")?.Value". If I update this Claim. When I read this from Claim, it was getting the previously Claim (I believe it is not refreshed). How do I refresh this Claim without using SignInManager/UserManager: 

var user = await _userManager.GetUserAsync(User);
await _signInManager.RefreshSignInAsync(user);

If I look at the database, the value is updated. If I logout and re-login, the value is updated. 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f75d9405-19a5-a14b-3a09-a3605b240750
* Version Independent ID: 1247ee64-fbc9-ce29-0710-f6e59380f5ee
* Content: [ASP.NET Core Blazor authentication and authorization](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/security/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/security/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
@HaoK ... Is the idea to inject a service into the component, the service has a method that uses UserManager/SignInManager (`RefreshSignInAsync`), and the user calls the service's method to accomplish a refresh of the user?

BTW and outside of the scope of this issue: Although we cover adding claims in Blazor WebAssembly, we don't discuss **_updating claims_**. However, I'll eventually reach the Blazor *Security & Identity* node of topics on my UE passes. I hope to identity and iron out missing coverage, such as this scenario, at that time.

**Comment by Alienroid** [comment]
I am guessing here, but I think the reason for the note "SignInManager<TUser> and UserManager<TUser> aren't supported in Razor components." is probably due to developers trying to call UserManager (which has a scoped lifetime) multiple times within same page. If a page contains 2(or more) components that both use UserManager, the app will crash due to "invalidOperationException: A second operation started on this context before a previous operation completed. Any instance members are not guaranteed to be thread-safe." 

Reference: https://github.com/dotnet/aspnetcore/issues/18340 and https://stackoverflow.com/questions/59747983/blazor-concurrency-problem-using-entity-framework-core (read the answer part)

Where as recommendation for DbContext is to use a factory or set as Transient, UserManager is just Scoped and we cannot change it to Transient.

TLDR; I wonder if we can still use UserManager, just need to make a note that  UserManager only works on page level and only one can be used within the page, else the app will crash.

**Comment by guardrex** [comment]
🐉 **_DRAGONS!_** 🐉 ... 😨🏃 ... :smile:

@Alienroid ... Did you consider a controller endpoint approach via ordinary Identity (RP/MVC) that refreshes the user?

**Comment by Alienroid** [comment]
Some context and I am thinking about another work around.

I am trying to add an user profile page, able to view, edit by the user and view by other users. I have a navigation bar that contains user profile image and notification (with outstanding notification counter). I have the issue of using UserManager twice, once in the page and one in the navigationbar; hence, I removed the NavBar call to UserManager for now.

I extended IdentityUser to be AppUser with added properties like FirstName, LastName, etc
I extended UserManagerExtenstion to call "await userManager.UpdateAsync(user);" to update user
Whenever I need to access user data, I'll call "await userManager.GetUserAsync(cpUser);"

With the page recommending UserManager is not supported, I've changed userManager.GetUserAsync(cpUser); to getting that data from " (await authenticationStateTask).User;"'s claims like: "user.Claims.FirstOrDefault(s => s.Type == "FirstName")?.Value"

I may move UserManagerExtension code to a service so I can update user (haven't tried this). But once that is updated, if I call "user.Claims.FirstOrDefault(s => s.Type == "FirstName")?.Value", it's still using the old data until I logout. Hence this open request.

I am just thinking about a workaround. Perhaps, I should create an user object that gets populated (from "user.Claims.FirstOrDefault..." and pass it via cascadingParameter/Value for the entire duration of the lifetime. Rest of the app won't be reading directly from "await userManager.GetUserAsync(cpUser);", but just read this user object. Whenever user relogin, that's the only time this user object will be populated directly from "await userManager.GetUserAsync(cpUser);". The disadvantage is that if the user data is outside of the scope (say from another browser), the user properties will be out of sync (which may be acceptable).

**Comment by Alienroid** [comment]
> 🐉 **_DRAGONS!_** 🐉 ... 😨🏃 ... 😄
> 
> @Alienroid ... Did you consider a controller endpoint approach via ordinary Identity (RP/MVC) that refreshes the user?

No. Do you mean after I update UpdateAsync, I'll redirect to a controller page that does 

var user = await _userManager.GetUserAsync(User);
await _signInManager.RefreshSignInAsync(user);

Then redirect back to the page?

That might work as well.

By the way, I don't get your emoticon. dragons ... shock... .run... laugh?

**Comment by HaoK** [comment]
Dragons are scary and randomly breathe fire, which is apt for trying to use User/SignInManager with blazor.  Yeah it was designed with very specific usage requirements in mind, as all of these threads/issues seem to be demonstrating

**Comment by guardrex** [comment]
😨 == **_fear mostly_** ... I'm worried that its easy to get into trouble and compromise security with a lack of official Blazor guidance that doesn't answer enough basic questions. I prefer that we have trustworthy, complete guidance for basic scenarios so that devs don't trust non-MS poor guidance (e.g., things found in random blog posts) ... or even just guess wrong and compromise security or have buggy apps.

> after I update UpdateAsync, I'll redirect to a controller page that does

Yes, exactly ... because the PU has said over and over (including those Javier posts) that the core Identity security approaches are all for Blazor **_but via controller Identity endpoints_**.

Hello @HaoK ... I leave this for you to advise me further on ... if something can/should be added. However, I'll hit these topics soon enough on UE passes, so we can do something quick/temporary with that NOTE if you want to.

**Comment by HaoK** [comment]
I'm not really sure what we should do here, calling @blowdart for thoughts, identity is pretty old and getting things to work outside of the controlled places like identity endpoints is definitely going to require some ninja skills...

My initial inclination would be to stick with the existing stance, where official advice is to only use identity inside of controllers, and everything else they are on their own.

**Comment by guardrex** [comment]
> only use identity inside of controllers

That's what I was thinking.

> The idea for having Blazor is to have seamless experience. If the profile page needs to load as controller, then it will work as in the past, but it's not as seamless.

That's been discussed fairly regularly IIRC on the PU repo.  I'm not aware of any other guidance from the PU other than use standard Identity [i.e., redirect the user to the Identity page/view + whatever controller(s) it uses]. Scaffold Identity into the app if custom processing is needed.

The other way is to hit an Identity controller from a component that manages the UI bits. I've done this in testing a while back, but I need to look up what the dev and I were discussing at the time. IIRC, it pertained to creating components that mirror what Identity pages/views do ... and I have a feel'in that the 🐉 might show up to roast and consume U! 👦🔥🥩🍽️ 

Let's leave this open because we can at least temporarily craft a sentence or two into the NOTE that says that ASP.NET Core Identity guidance should be followed:

* Use Identity pages/views/controllers (scaffold-in if needed for custom processing) + cross-link. This seems safe and recommended. The NOTE won't leave the dev hanging if we add something like this.
* Call directly to Identity controller endpoints from a component. This has to be vetted by the PU ... is it safe? is it stable?

**Comment by guardrex** [comment]
Yeah ... I did put the styling guidance into the scaffolding Identity topic at ...

https://docs.microsoft.com/en-us/aspnet/core/security/authentication/scaffold-identity?view=aspnetcore-5.0&tabs=visual-studio#style-authentication-endpoints

... and AFAICT, all of the Blazor guidance for Identity is for scaffolding Identity/using Identity, so the NOTE can just link to the scaffolding information. Such cross-links exist in the Blazor Server *Overview* at ...

https://docs.microsoft.com/en-us/aspnet/core/blazor/security/server/?view=aspnetcore-5.0&tabs=visual-studio#scaffold-identity

**Comment by Alienroid** [comment]
I actually implemented most of the Scaffolded Identity into razor components. There are few things that cannot be done like login as cache can't be accessed by javascript/wasm (can't' call PasswordSignInAsync / https://github.com/aspnet/AspNetCore/issues/15720). But lots of features are actually quite safe to be done. Like download personal data, delete account, change password (need redirect), change email (need redirect), even register (need redirect). In fact, only login/relogin and logoff/logoff redirect cannot be done.

Blazor is awesome in the sense it adds additional layer in terms of security. Just because few cases (like login/logout) doesn't work, doesn't mean rest of the identity can't enjoy blazor's benefits. :)

**Comment by Alienroid** [comment]
FYI I just added this to scaffold identity folder that I think should help.

RefreshSignIn.cshtml

```
@page
@using Microsoft.AspNetCore.Identity
@attribute [IgnoreAntiforgeryToken]
@inject SignInManager<AppUser> SignInManager
@inject UserManager<AppUser> UserManager
@functions {
    public async Task<IActionResult> OnPost()
    {
        if (SignInManager.IsSignedIn(User))
        {
            var user = await UserManager.GetUserAsync(User);
            await SignInManager.RefreshSignInAsync(user);
        }

        return Redirect("~/");
    }
}

```

**Comment by guardrex** [comment]
Yes, exactly ... I was hoping that that approach :point_up: would work. You can pass the XSRF token to components and then not ignore it ...

https://docs.microsoft.com/aspnet/core/security/authentication/scaffold-identity#pass-an-xsrf-token-to-the-app

**Comment by Alienroid** [comment]
Got it to work.

Just curious, I am on server side, not sure why you brought up Xsrf as an example. If there's any altercation, the signalR connection would be interrupted and disconnected. 

Are you referring to above [IgnoreAntiforgeryToken] because I got that from logoff (from sample)

**Comment by guardrex** [comment]
Yes, I speculate that `[IgnoreAntiforgeryToken]` can be removed if the token is attached to the request. I see the PU discussing how `[IgnoreAntiforgeryToken]` must be there for Blazor Server per scaffolding over the default template Logoff with one that doesn't have `[IgnoreAntiforgeryToken]` ...

* https://github.com/dotnet/Scaffolding/issues/1287
* https://github.com/dotnet/Scaffolding/issues/1423

... because the token isn't present for any component POST to Identity pages/controllers unless its explicitly made available to Blazor via the guidance that I cross-linked.

**Comment by Ongax05** [comment]
> FYI I just added this to scaffold identity folder that I think should help.
> 
> RefreshSignIn.cshtml
> 
> ```
> @page
> @using Microsoft.AspNetCore.Identity
> @attribute [IgnoreAntiforgeryToken]
> @inject SignInManager<AppUser> SignInManager
> @inject UserManager<AppUser> UserManager
> @functions {
>     public async Task<IActionResult> OnPostAsync(string returnUrl = null)
>     {
>         returnUrl = returnUrl ?? Url.Content("~/");
>         if (SignInManager.IsSignedIn(User))
>         {
>             var user = await UserManager.GetUserAsync(User);
>             await SignInManager.RefreshSignInAsync(user);
>         }
> 
>         return Redirect(returnUrl );
>     }
> }
> ```

I have been trying for a week to refresh the blazor session and I had already tried this but it gave me an error with the AntiforgeryToken and I have to thank you for this.
